starknet_devnet/transactions.py

```diff
@@ -1,11 +1,150 @@
-from .constants import FAILURE_REASON_KEY
 
+class DevnetTransaction:
+    """Represents the devnet transaction"""
+
+    def __init__(
+        self,
+        internal_tx: InternalTransaction,
+        status: TransactionStatus,
+        execution_info: TransactionExecutionInfo or StarknetTransactionExecutionInfo,
+        transaction_hash: int = None,
+    ):
+        self.block = None
+        self.execution_info = execution_info
+        self.internal_tx = internal_tx
+        self.status = status
+        self.transaction_failure_reason = None
+        self.transaction_index = 0
+        self.transaction_hash = transaction_hash
+
+        if transaction_hash is None:
+            self.transaction_hash = internal_tx.hash_value
+
+    def __get_actual_fee(self) -> int:
+        """Returns the actual fee"""
+        return self.execution_info.actual_fee if hasattr(self.execution_info, "actual_fee") else 0
+
+    def __get_events(self) -> List[Event]:
+        """Returns the events"""
+        contract_address = self.execution_info.call_info.contract_address
+        return [Event.create(event_content=e, emitting_contract_address=contract_address) for e in self.execution_info.call_info.events]
+
+    def __get_l2_to_l1_messages(self) -> List[L2ToL1Message]:
+        """Returns the l2 to l1 messages"""
+        l2_to_l1_messages = []
+
+        if not hasattr(self.execution_info.call_info, "l2_to_l1_messages"):
+            return l2_to_l1_messages
+
+        contract_address = self.execution_info.call_info.contract_address
+
+        for l2_to_l1_message in self.execution_info.call_info.l2_to_l1_messages:
+            l2_to_l1_messages.append(L2ToL1Message(
+                from_address=contract_address,
+                to_address=Web3.toChecksumAddress(hex(l2_to_l1_message.to_address)),
+                payload=l2_to_l1_message.payload,
+            ))
+
+        return l2_to_l1_messages
+
+    def __get_block_hash(self) -> int:
+        """Returns the block hash"""
+        return self.block.block_hash if self.block else None
+
+    def __get_block_number(self) -> int:
+        """Returns the block number"""
+        return self.block.block_number if self.block else None
+
+    def set_block(self, block: StarknetBlock):
+        """Sets the block hash and number of the transaction"""
+        self.block = block
+
+    def set_failure_reason(self, error_message: str):
+        """Sets the failure reason of the transaction"""
+        # set tx_id?
+        self.transaction_failure_reason = TransactionFailureReason(
+            code=StarknetErrorCode.TRANSACTION_FAILED.name,
+            error_message=error_message
+        )
+
+    def get_signature(self) -> List[int]:
+        """Returns the signature"""
+        return self.internal_tx.signature if hasattr(self.internal_tx, "signature") else []
+
+    def get_tx_info(self) -> TransactionInfo:
+        """Returns the transaction info"""
+        return TransactionInfo.create(
```

Comment:
If performance is an issue, maybe have this cached, `if not self.tx_info: self.tx_info = ...; ... return self.tx_info`. Same for other returned objects (TransactionTrace, TransactionExecution).

Reply:
Perfomance is not the issue, it was because deployment tx hash was calculated twice.

We don't want to cache receipt and info as they change depending on the block information. Adding updates and caching would introduce unnecessary overhead IMHO and would proved minimal benefit.

Reply:
Ok, and now the double calculation issue is resolved?

Reply:
Correct